api/v1/odhdeployment_types.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ManifestType represents the type of the manifest source
+type ManifestType string
+
+const (
+	// HelmManifestType represents a Helm chart source
+	HelmManifestType ManifestType = "helm"
+
+	// KustomizeManifestType represents a Kustomize source
+	KustomizeManifestType ManifestType = "kustomize"
+)
+
+// OdhDeploymentSpec defines the desired state of OdhDeployment
+type OdhDeploymentSpec struct {
+	Components []Component `json:"components,omitempty"`
+	Version    string      `json:"version,omitempty"`
+}
+
+type Component struct {
+	Name              string            `json:"name"`
+	URL               string            `json:"url"`
+	Type              ManifestType      `json:"type"`
+	ManifestFolder    string            `json:"manifestFolder,omitempty"`
+	CustomParameters  map[string]string `json:"customParameters,omitempty"`
+	PostInstallScript string            `json:"preInstallScript,omitempty"`
+	PreInstallScript  string            `json:"postInstallScript,omitempty"`
+}
```

Comment:
> Have Component be a tagged union (or however we should do that in Go)
So we would define an interface, and then have kustomize/helm types that implement the interface? Seems reasonable.
> Have two distinct CRD
I think this might be the better option of the two. Helm and kustomize installs may even require us to specify different types of parameters for them, so separate CRDs would allow users to better customize their deployments, right?

Reply:
Yes, I would favor that solution as well.
We would have two controllers (conceptually, they'd live in the same executable) with each their control-loop.

Reply:
I have updated the code to use custom plugins. Given that we no longer need the `Type` field and can just use Helm Plugin for the helm charts

Reply:
What does it mean for helm charts concretely ? Can the OdhDeployment resource simply points to an helm chart url or does it have to be a Kustomization itself referencing an helm chart (aka = one more level of indirection) ?